Aten vector default constructors set to 0, add fnmadd and fnmsub

sharpobject opened the pull request:
Test Plan: The only functional change is zero-initialization instead of undefined-initialization. If tests pass, I think it should be fine.

Differential Revision: D75345074




cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168

sharpobject commented:
Previously this was https://github.com/pytorch/pytorch/pull/154060

Apologies for the mess.

I've added new tests in vec_test_all_types.cpp, but they were immediately failing for c10::Half. I discovered that the tests for fmadd are also failing for the same reason, but we've ignored their constant failure internally. Basically, in the worst case (and speaking of fmsub here, for the sake of the example, but this applies to all 4 variations of fma), the test comes up with a,b,c such that a*b is approximately equal to c, then a*b-c is an example of [catastrophic cancelation](https://en.wikipedia.org/wiki/Catastrophic_cancellation). So as it is computed at low precision it has a lot of error compared to if we computed it with much higher precision. And this is precisely what the test does, it just converts the operands to doubles and evaluates the same expression to get the expected result. Additionally, c10::Half was probably receiving the default error bounds of 0 in the tests. So I've changed the error bound to 1e-2, that is to say 1%, but in most real world usage the error should be a lot less than this. We only need it to be so large to pass the numerics test in the case where we've shoved catastrophic cancelation into a single instruction 😭😭😭

After this change, we should be able to stop ignoring the existing tests for fmadd internally. 

cc @swolchok

swolchok reviewed (changes requested):
looks good except for copy/pasted code

sharpobject committed:
Aten vector default constructors set to 0, add fnmadd and fnmsub (#154298)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/154298

Test Plan: The only functional change is zero-initialization instead of undefined-initialization. If tests pass, I think it should be fine.

Differential Revision: D75345074

sharpobject commented:
Currently this is failing one test in Github Actions with

```
REGRESSION: benchmark ('float_args', 'compile_time_instruction_count') failed, actual result 461490130 is 1.99% higher than expected 452500000 ±+1.50% if this is an expected regression, please update the expected results.
```

Which seems unrelated to the changes. @swolchok are we able to proceed or do we need this to pass? Thanks.

swolchok reviewed (approved):
copy/pasted code is fixed, accepting

izaitsevfb commented:
@pytorchbot revert -m "breaks linux-jammy-aarch64-py3.10 / build" -c ignoredsignal

[GH job link](https://github.com/pytorch/pytorch/actions/runs/15358751590/job/43222748516) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/489afa829a248ca64c4b2dffe2e6d601b8816cf9)
